vignettes/grouping.Rmd

```diff
@@ -0,0 +1,172 @@
+  mutate(batch = sample(c("a", "b"), nrow(.), replace = TRUE)) %>%
+  select(-c(VF, F, M, L))
+
+hpc
+```
+
+If we wanted to compute the accuracy of the first resampled model, we could write:
+
+```{r acc-1}
+hpc %>% 
+  filter(Resample == "Fold01") %>%
+  accuracy(obs, pred)
+```
+
+The metric function returns one row, giving the `.metric`, `.estimator`, and `.estimate` for the whole data set it is passed.
+
+If we instead group the data by fold, metric functions like `accuracy` will know to compute values for each group; in the input, each row will correspond to a Resample.
+
+```{r hpc-cv-2}
+hpc %>% 
+  group_by(Resample) %>%
+  accuracy(obs, pred)
+```
+
+Note that the first row, corresponding to `Resample`, gives the same value as manually filtering for the observations corresponding to the first resample and then computing the accuracy.
+
+This behavior is what we mean by group-awareness. When passed grouped data, metric functions will return values for each group.
+
+# Group-wise metrics
+
+Group-wise metrics are associated with a data-column such that, when passed data with that column, the metric will temporarily additionally group by that column, compute values for each of the groups defined by the column, and then aggregate the values computed for the temporary grouping back to the level of the inputted data's grouping.
+
+More concretely, let's turn to an example where there is no pre-existing grouping in the data. Consider the portion of the HPC data pertaining to the first resample:
+
+```{r res-1}
+hpc %>% 
+  filter(Resample == "Fold01")
+```
+
+Suppose that the `batch`es in the data represent two groups for which model performance ought not to differ. To quantify the degree to which model performance differs for these two groups, we could compute accuracy values for either group separately, and then take their difference. First, computing accuracies:
+
+```{r acc-by-group}
+acc_by_group <- 
+  hpc %>% 
+  filter(Resample == "Fold01") %>%
+  group_by(batch) %>%
+  accuracy(obs, pred)
+
+acc_by_group
+```
+
+Now, taking the difference:
+
+```{r diff-acc}
+diff(c(acc_by_group$.estimate[2], acc_by_group$.estimate[1]))
+```
+
+Group-wise metrics encode the `group_by()` and subtraction step shown above into a yardstick metric. We can define a new group-wise metric with the `new_groupwise_metric()` function:
+
+```{r}
+accuracy_diff <-
+  new_groupwise_metric(
+    fn = accuracy,
+    name = "accuracy_diff",
+    aggregate = function(acc_by_group) {
+      diff(c(acc_by_group$.estimate[2], acc_by_group$.estimate[1]))
+    }
+  )
+```
+
+* The `fn` argument is the yardstick metric that will be computed for each data group.
+* The `name` argument gives the name of the new metric we've created; we'll call ours "accuracy difference."
+* The `aggregate` argument is a function defining how to go from `fn` output by group to a single numeric.
+
+The output, `accuracy_diff`, is a function subclass called a `metric_factory`:
```

Comment:
Trying to talk about functions as objects while adhering to tidyverse style is a bit difficult here. I'm game for however yall want to approach this. :)